api/v6/api.go

```diff
@@ -16,9 +16,23 @@ type ImageStatus struct {
 	Containers []Container
 }
 
+// ReadOnlyReason enumerates the reasons that a controller is
+// considered read-only. The zero value is considered "OK", since the
+// zero value is what prior versions of the daemon will effectively
+// send.
+type ReadOnlyReason string
+
+const (
+	ReadOnlyOK      ReadOnlyReason = ""
+	ReadOnlyMissing ReadOnlyReason = "NotInRepo"
+	ReadOnlySystem  ReadOnlyReason = "System"
+	ReadOnlyNoRepo  ReadOnlyReason = "NoRepo"
+)
+
 type ControllerStatus struct {
 	ID         flux.ResourceID
 	Containers []Container
+	ReadOnly   ReadOnlyReason `json:",omitempty"`
```

Comment:
What's the motivation behind making this `omitempty` but none of the other properties?

It's plausible that a `Controller` could have no `Policies`, or maybe even no `Containers`, just as it may not be `ReadOnly`.

Reply:
There's no especial reason, and in fact I think this is the wrong way around for this particular field -- I _want_ a missing value (e.g., from an old daemon) to look like an empty string, for API consumers.

Reply:
With or without `omitempty`, unmarshalling with a missing value will result in an empty string in the consumer. Only way to distinguish missing versus empty is to make this field a pointer.

I suppose if we care about saving a few bytes we can pepper our structs with `omitempty` and hope any other languages consuming our api Do The Right Thing.

Reply:
> unmarshalling with a missing value will result in an empty string in the consumer.

From Go, yes. But not from say, JavaScript.